### ISSUE TITLE
Excluding certain fields when calculating conditions-hash

### ISSUE THREAD

**Issue by kamil-olszewski-uw** [erasmus-without-paper/ewp-specs-api-iias#48]
There are fields in the cooperation coditions that are not in the IIAs template. Those are:

- sending-contact
- receiving-contact
- sending-ounit-id
- receiving-ounit-id

Recently, there has been the idea that certain fields should be removed from the conditions-hash computations, because changing these fields would change the hash, but they do not contain information that should affect the validity of the agreement.

We believe that this should apply to fields sending-contact and receiving-contact, as the contact details of cooperation coordinators can indeed change.

However, we believe that the hash should take into account the field sending-ounit-id and receiving-ounit-id, because the faculties implementing cooperation conditions are unlikely to change. And if this happens (or the name of the organizational unit changes due to changes in the structure of the university), a new agreement should be signed in such a case.

Please comment if you think we should do it differently.

**Comment by fmapeixoto** [comment]
In my opinion, the contacts shouldn't be calculated because, as referred by Kamil, "...do not contain information that should affect the validity of the agreement.".

Regarding the OUnits in the cooperation conditions, I have 2 use cases that I understand may exists or not:
1. in case a OUnit is specified at the partners level, no more different OUnits should appear in the cooperation conditions right? if affirmative they can:
  a)  be present (required)
  b) present (optional)
  c) none at all (because they are already at the top
2. in case no OUnit is specified at the partner level, then the possibility of having OUnits in the Cooperation Conditions is something that may be needed in much cases.
In order to make easier for specifications, the optional parameters may be the best option.

**Comment by kamil-olszewski-uw** [comment]
> 1. in case a OUnit is specified at the partners level, no more different OUnits should appear in the cooperation conditions right? if affirmative they can:
>    a)  be present (required)
>    b) present (optional)
>    c) none at all (because they are already at the top

There may be situations when e.g.:

- faculty is specified at the partner level, and institutes are specified at the cooperation conditions level,
- subsidiary of the university is specified at the partner level, and its faculties are specified at the cooperation conditions level.

Of course, an erroneous situation may arise when, for example, faculty is specified at the partner level, and other faculties are specified at the cooperation conditions level. But it is a business error that should be noticed by the user and not by the system.

**Comment by fmapeixoto** [comment]
And one more remark, does it make sense, per cooperation condition, to have it 0-unbounded? Shouldn't it be just 0-1 per cooperation? Just asking because I don't know practical use-cases of these.

**Comment by umesh-qs** [comment]
Restricting to a maximum of 1 can also resolve https://github.com/erasmus-without-paper/ewp-specs-api-iias/issues/29. This was closed today without any solution.

**Comment by EvelienRenders** [comment]
I don't know if this is relevant, but the Erasmus Dashboard right now only allows one OUnit per Agreement (so not even on the cooperation level). Is this something that EUF should 'fix'? 

For example: @umesh-qs, it might prove weird when one of your MoveOn users starts an agreement with 4 cooperation conditions that each has a different OUnit, and it comes into the Dashboard as 4 separate agreements. 

What happens if the Dashboard user then changes things on an agreement level, and sends it back? What does the MoveOn user see then?

As for your remark @fmapeixoto 
> And one more remark, does it make sense, per cooperation condition, to have it 0-unbounded? Shouldn't it be just 0-1 per cooperation? Just asking because I don't know practical use-cases of these.

It depends; right now the Dashboard users are making up OUnits for 'institution-wide' agreements - because they cannot choose multiple OUnits per Agreement in the Dashboard. That means that we have 7 faculties + 1 institution-wide OUnit. 

That is not a problem per se, but might not be the perfect way to use these Units...

**Comment by umesh-qs** [comment]
@EvelienRenders 
I did not understand whether you are supporting multiple cooperation conditions or not. But it is certainly possible.
How it is implemented in the Dashboard is something that is specific to Dashbord and if it is implemented the way you described then personally that is not the best way.

But that is not what is there under the issue https://github.com/erasmus-without-paper/ewp-specs-api-iias/issues/29. What I was asking is that when there is no limit on the number of cooperation conditions then how do I uniquely identify each cooperation condition. Restriction to max 1 for each type can resolve the issue.

**Comment by EvelienRenders** [comment]
@umesh-qs
Forgive my non-technical question ;) I'm an information manager at a Dutch university and not working directly on the development of the connection to our SIS (Osiris, developed by CACI). I'm trying to get a grip of the issues that are still ongoing - and I understand your issue #29 - although unfortunately, I can only add an HEI's perspective on these issues. 

I find it comforting to hear that you think the Dashboard's implementation is not the best because personally, I'm thinking that as well - I will send EUF the feedback via their [support portal](https://support.uni-foundation.eu/) once it is online again... 

Thank you for all the hard work that is going into this on your end, but also on the ends of the other 3rd party providers - it's much appreciated (although not many might let you know during the 'trial and error' phase 😉)

**Comment by pleys** [comment]
@kamil-olszewski-uw I think we should leave contacts out of conditions hash. I see a great advantage of having contacts connected the the agreements, e.g. for sharing the academic coordinator of a given agreements (who can be the responsible person for signing learning agreements) but this person can change every year and we don't want to re-approve our agreement every year.

**Comment by pleys** [comment]
The organisational units discussion is somewhat more complex and the discussion above points to the issue of grouping agreements that I raised earlier.

**Comment by umesh-qs** [comment]
@kamil-olszewski-uw 
I am sorry, I did not understand, what are we trying to solve here by adding/removing fields from the calculation of hash?

**Comment by kamil-olszewski-uw** [comment]
> I am sorry, I did not understand, what are we trying to solve here by adding/removing fields from the calculation of hash?

The point is that the mobility coordinators are not part of the agreement. They would not be on a traditional paper agreement. However, since the coordinators are part of the cooperation conditions in API, the change of the coordinator would make the hash, which was the basis for accepting the agreement, obsolete. And the change of coordinator is something common. Therefore, we want sending-contact and receiving-contact not to be taken into account when hash is calculated.

**Comment by fmapeixoto** [comment]
I believe the discussion here should be directly related to whether we remove some fields from the calculation of the Hash, and or they should exist, and if so, in which circumstances.

For the IIA ecosystem (all APIs) to work properly, all the systems (ownsys and providers) have to guarantee that they are able to comply with all the fields (at least for the calculation of the hash) and have proper ways of storing and sending the exact same values, in the same order, so the hash match every time. I don't think we should discuss here already what is supported by whom because what is agreed here should be implemented to comply.

I agree with Paul regarding the contacts. Regarding the ounits, I leave the he discussion to the ones who have more deep knowledge about all the use-cases and needs.

**Comment by umesh-qs** [comment]
@kamil-olszewski-uw 
I think too much weightage is given to hash. Your scenario is frequent changes to agreement contact. Others are talking about ounit.
I have a problem that recent API changes will change the hash of all the agreements exchanged via EWP network.
Are our clients supposed to approve all of them again now and again in future if more changes are done?

**Comment by fmapeixoto** [comment]
@umesh-qs that is actually a good question which I believe is related because what else is proof that a specific agreement has been "signed" if the hashes don't match on both sides? Perhaps coming back to the topic of having a digital signature in the XML somehow fixes everything. But is the same of having the fields and not being able to change any of them after "signing" (including contacts).

**Comment by kamil-olszewski-uw** [comment]
> I have a problem that recent API changes will change the hash of all the agreements exchanged via EWP network.
> Are our clients supposed to approve all of them again now and again in future if more changes are done?

We are currently in the infant stage of the IIAs Approval API. I cannot speak for others, but we do not have any digitally approved IIA yet, not including test agreements. And even if some agreements between partners using EWP have been approved via EWP, there should still be signed agreements on paper. The most important thing is to meet the specification on the deadlines set by the EC regarding agreements, which will be signed only digitally.

**Comment by kamil-olszewski-uw** [comment]
There are a few important issues in this thread - related, but still separate, so we have a slight chaos here. I will try to answer here to unanswered questions:

**1. Number of organizational units with cooperation conditions:**

We understand that @fmapeixoto asked about the 0-unbounded to 0-1 change in the context of organizational units under cooperation conditions, and not in the context of the cooperation conditions themselves. We don't remember who postulated the unbounded organizational units and think we can limit them to 0-1 as long as no one objects.

**2. Number of cooperation conditions of one kind:**

We cannot change cooperation conditions from 0-unbounded to 0-1. A simple use case where the unbounded option matters:

- Student mobilities for studies from university A to university B, EQF level 6: 5 people, 25 months
- Student mobilities for studies from university B to university A, EQF level 6: 4 people, 20 months
- Student mobilities for studies from university A to university B, EQF level 8: 3 people, 16 months

**3. Differences between Dashboard and API:**

@EvelienRenders - If something in the Dashboard currently looks different than the final API shape that we collectively agree to, we are convinced that the Dashboard will be adapted to the API structure.

**Comment by kamil-olszewski-uw** [comment]
**SUMMARY (our proposed changes to IIAs API based on our discussion):**

1. We will remove sending-contact and receiving-contact from the calculation of the hash.

2. We will NOT remove sending-ounit-id and receiving-ounit-id from the calculation of the hash.

3. We will change maxOccurs of sending-ounit-id and receiving-ounit-id from unbounded to 1.

4. We will NOT change max Occurs of student-studies-mobility-spec, student-traineeship-mobility-spec, staff-teacher-mobility-spec and staff-training-mobility-spec from unbounded to 1.

5. We will NOT force a specific number of digits in the isced-f-code, but in the commentary to this element we will strongly recommend that the ISCED code should be four-digit, because this way it will be compliant with the requirements of Mobility Tool+ apllication, which only accepts four-digit codes for mobilities (see issue:   #49).

**Comment by umesh-qs** [comment]
Please suggest how to handle hash/approval for existing agreements in the EWP network.

**Comment by umesh-qs** [comment]
@kamil-olszewski-uw 
Not sure if this was discussed before releasing the new IIA API. How do we handle hash changes for the approved agreements in EWP?

**Comment by EvelienRenders** [comment]
With regards to point 3 - we found that this would be a bit restrictive. Could someone explain elaborate if this is a technical requirement?

> 3. We will change maxOccurs of sending-ounit-id and receiving-ounit-id from unbounded to 1.

**Comment by umesh-qs** [comment]
@kamil-olszewski-uw 
I think IIA API changes have been released, without considering all the points. I would like to know what is the process for approving the changes for an API and who takes the final call to release the changes even if there are unanswered questions?

**Comment by kamil-olszewski-uw** [comment]
> > We will change maxOccurs of sending-ounit-id and receiving-ounit-id from unbounded to 1.

> With regards to point 3 - we found that this would be a bit restrictive. Could someone explain elaborate if this is a technical requirement?

Do you really have individual cooperation conditions dedicated to several (but not all) organizational units? We have several thousand agreements with universities from the EU and almost as much with the rest of the world, and it was always enough for us to indicate one unit or no unit at all.

**Comment by kamil-olszewski-uw** [comment]
> Not sure if this was discussed before releasing the new IIA API. How do we handle hash changes for the approved agreements in EWP?

Do you mean (1) changes to the existing agreement due to the modification of the cooperation conditions or (2) obsolescence of the hashes calculated so far due to the change in the method of calculating the hash?

Ad 1. We thought about it and many voices were in favor of the solution that the change in cooperation conditions should result in signing a new agreement.

Ad 2. As I wrote earlier, we are currently in the infant stage of the IIAs Approval API. First stable version was released few years ago. It is difficult to consider earlier versions as an unchanging basis without the risk of numerous backwards incompatible changes

**Comment by umesh-qs** [comment]
@kamil-olszewski-uw 
So this is an official stand that because of new changes to the API structure all the existing agreements are not valid anymore until approved again? Is it ok to communicate this to our clients once we release the new changes?

**Comment by umesh-qs** [comment]
@kamil-olszewski-uw 
Awaiting your response to my last comment, please.

**Comment by janinamincer-daszkiewicz** [comment]
This is not true that because of the new changes to the API structure *all* the existing agreements are not valid anymore.
First of all they have already been approved. No need to approve them again.
Second, they will stay approved under the old API.
It is natural that the new APIs will show up and will have to be implemented.
We have nothing else to add so please do not push.

**Comment by umesh-qs** [comment]
"We have nothing else to add so please do not push.". This is a very harsh comment.
I was just trying to get some answers since I have to respond to multiple stakeholders in my organization.
And I don't have the option to tell them not to push me, since EWP API designers/approvers don't like asking questions.

Let me come back to the issue of hash change, with an example
1. HEI A creates and shares an IIA with HEI B.
2. HEI B pulls the IIA data.
3. HEI B validates the hash and approves the IIA of HEI A.

Now both HEI A and HEI B implements IIA API version 6.0

4. HEI B pulls the data of the same IIA from HEI A.
5. Since there is a change in hash calculation logic in the new API, HEI B will get a new hash and has to approve again.

Can any of the IIA service provider, suggest ways, where our clients, don't have to re-approve the already approved agreement in the above scenario?

**Comment by sascoms** [comment]
If I have not misunderstood the question:

There two ways:

1. To keep the approval status in your systems. 
This way, even if the hash or the api version changes, you still have a previous agreement set as approved in the past. 
And no need for a re-approval operation.
Or even if the API version changes, these IIAs are already approved. No need to re-approve (if there is no IIA content updated/changed)

2. To send CNR to the partner and re-initiate the approval process.  

In my opinion, already approved IIAs do not need to go through a new approval check or process as the re-approval process should apply only if one of the partners change any of the content or the valid years or similar data in the IIA.

However, if the IIA API version sets new required fields and these fields are missing in the existing IIAs and the provider systems require these fields to exist for approval process or mobilities, then #2 option should definitely be applied.

**Comment by fmapeixoto** [comment]
This will be a problem over time, not only from 5 to 6 or 4 to 5.
In my opinion, adding the signature field in the XML will make it irrefutable and, if signed by both, APIs can change the whole structure that the signature of the saved XML is always valid and you can keep it in the DB. Instead of calculated hashes, signatures of both parties are enough.
The past ones, we have to trust that partners support older versions, or at least keep the older hashes in the approvals API.

**Comment by kamil-olszewski-uw** [comment]
As Francisco wrote, this problem may reappear many times in the future - for example, when the EC decides to introduce completely new elements to the cooperation conditions.

We plan to add signatures in the future. But this is a more general issue that also applies to other APIs.

Currently, for example, you can use such a solution: If the institution wants the approve agreement only once - with a hash that will never expire and can always be used to check whether the agreement has not changed - then the institution may keep an information in the system, which the IIAs API version was used during approval of the agreement, and the system may have hash calculation algorithms implemented for different versions of the IIAs API.

**Comment by fmapeixoto** [comment]
If, for already approved IIAs, we just check the approvals API for the partner hash and our DB, they should match, we don't need to recalculate it.

**Comment by kamil-olszewski-uw** [comment]
> If, for already approved IIAs, we just check the approvals API for the partner hash and our DB, they should match, we don't need to recalculate it.

I'm afraid I didn't understand your idea :-) Could you please present it as a simple scenario?

**Comment by fmapeixoto** [comment]
The API IIAs Approval have the following entries:
approval>
    iia-id>82a94c1684ce-947c-7ac9-cd7f-287f065a/iia-id
    conditions-hash>3d3310935590256ecff21549b45eb4e1662b0f2a2aafcb7d91ea5f271f4d9e07/conditions-hash>
/approval>
When you retrieve a partner IIA (with IIA Get):
- The sending partner can check the receiving partner Approval API (or it's own) for the ID and don't need to recalculate the hash
- Upon receiving, the receiving partner can check the sending partner Approval API and compare with it's own DB.

If they match there is no need to recalculate hash of a particular IIA as it is already approved, independent of the version, right?

**Comment by kamil-olszewski-uw** [comment]
What exactly would a partner's hash be compared to in "it's own DB"? To a previously received (during approving) partner hash, which we keep in our DB? The positive result of the hash comparison will not be any proof that the partner has not changed anything in the cooperation conditions in the meantime. After all, a partner can send cooperation conditions with an incorrect hash. This is why it is necessary to calculate the hash during approving and therefore it would be necessary later if we would like to check if there were any illegal changes after the approval.

**Comment by EvelienRenders** [comment]
> > > We will change maxOccurs of sending-ounit-id and receiving-ounit-id from unbounded to 1.
> 
> > With regards to point 3 - we found that this would be a bit restrictive. Could someone explain elaborate if this is a technical requirement?
> 
> Do you really have individual cooperation conditions dedicated to several (but not all) organizational units? We have several thousand agreements with universities from the EU and almost as much with the rest of the world, and it was always enough for us to indicate one unit or no unit at all.

Thanks for the question Kamil!

We do have so-called Joined Faculty Agreements - where we normally put multiple faculties on one condition. However, I've discussed this internally, and it would not be a problem for us to separate them over two cooperation conditions in the future - one for each OUnit (faculty).

**Comment by frangarcj** [comment]
Trying to not be too much off-topic, I have some questions arising from this discussion:

1. Is there any recommendation on what api version (client/server) to maintain? how many?
2. Is there any roadmap on api version increments? 
3. Also a lot of HEIs are in the dashboard, do you know if they have a roadmap for api versions?

We were finishing our IIAs 3.0 implementation in our tool at the University of Almería and suddenly versions 4, 5 and 6 appeared, so we are a little confused.

Thanks in advance

**Comment by janinamincer-daszkiewicz** [comment]
> Is there any recommendation on what api version (client/server) to maintain? how many?

The most recent. As many as it makes sense. If all your partners have upgraded to the newer versions there is not needed to keep suport for the older ones.

>  Is there any roadmap on api version increments?

We try not to increment version without specific need. Recent changes were  initiated by the new templates of LA and IIA announced by the commision.

>   Also a lot of HEIs are in the dashboard, do you know if they have a roadmap for api versions?

I am sure that the developers of the Dashboard and OLA will upgrade to the most recent versions as soon as possible. I can not give any dates on their behalf.

**Comment by umesh-qs** [comment]
@janinamincer-daszkiewicz 
1. Yes it makes sense to keep the most recent version of API. But, is there a cutoff date by when everyone in EWP network has to move to the latest version?
2. Are all the recent changes to the API, result of the new announcement by the commission?

**Comment by janinamincer-daszkiewicz** [comment]
> Yes it makes sense to keep the most recent version of API. But, is there a cutoff date by when everyone in EWP network has to move to the latest version?

I am not in the authority to decide about such dates.

> Are all the recent changes to the API, result of the new announcement by the commission?

Some are the results from other discussions, in Github, by e-mails. We ask developers to put the information in Github, to make it public.

**Comment by umesh-qs** [comment]
Ok, so who has the authority to decide on putting the cut off date? Will contact them.

On the changes based on Github discussions, emails ... who takes a call on that and what is the basis of it?

**Comment by janinamincer-daszkiewicz** [comment]
I don't know, may be DGEAC?

**Comment by umesh-qs** [comment]
@janinamincer-daszkiewicz 
1. Any specific person/team in DGEAC with whom the EWP API team is working for the requirement?
2. On the changes based on Github discussions, emails ... who takes a call on that and what is the basis of it?
3. Who all are part of the API design/release team? I assume at the moment that is it you and Kamil. Anyone else?

**Comment by jpbacelar** [comment]
Umesh, if you have questions about the EWP governance feel free to contact me. The purpose of this git is to provide a platform to carry out technical discussions with the EWP community, and I sense your current interests lie elsewhere. 

As for decision-making we are very lucky to have a team that spans several universities, and that has a pretty decent track record when it comes to take into consideration not just technical input (e.g. from the git) but also policy and user requirements. Of course your question is disingenuous, since you met most of these colleagues when attending EWP meetings...

**Comment by umesh-qs** [comment]
Hi Joao,
Any questions raised or arguments posted here are specific to this topic and should be seen in that context. Let us stick to this topic and not divert by bringing in the credibility question. I don't think anyone here is questioning anyone's credibility, including me.

My interest lies with my organization and the clients using our product. 
If I am not getting proper answers, or getting rude/intimidating answers ("I don't know, may be DGEAC?", "We have nothing else to add so please do not push"), then yes, you might have to be involved and maybe frequently. And all I was looking at to find the right answers. 

I was looking for answers to technical issues here. Other questions came up because of the responses received. I am happy to contact you directly for any concerns related to any API, if that is what seems right.

I still don't have a 100% working solution on how to handle approved agreements on the hash change when different service providers are using a different version of the same API.
If there is no 100% solution and API changes are mandatory, then I feel, that has to be communicated accordingly. This will help us to communicate further to our clients so that they are aware of the impact of changes on their workflow.

**Comment by jpbacelar** [comment]
Hi Umesh,

Glad to hear we can get back on track, thanks for that. I can't solve the hash issue myself, but I have picked up several topics raised on this thread for the next project meeting - not least the matter of deprecating APIs, which is undoubtably pertinent. Hope we can borrow from this discussion to provide greater clarity and a more consistent framework for the dev teams.

**Comment by umesh-qs** [comment]
Hi,
Can someone help me with hash calculation logic?
For the below content. What should be the hash?
```
<ns40:cooperation-conditions>
<ns40:student-traineeship-mobility-spec>
<ns40:sending-hei-id>uni1</ns40:sending-hei-id>
<ns40:receiving-hei-id>uni2</ns40:receiving-hei-id>
<ns40:receiving-academic-year-id>2018/2019</ns40:receiving-academic-year-id>
<ns40:mobilities-per-year>1</ns40:mobilities-per-year>
<ns40:recommended-language-skill>
<ns40:language>en</ns40:language>
<ns40:cefr-level>C1</ns40:cefr-level>
</ns40:recommended-language-skill>
<ns40:isced-f-code>0313</ns40:isced-f-code>
<ns40:total-months>2</ns40:total-months>
<ns40:blended>false</ns40:blended>
</ns40:student-traineeship-mobility-spec>
</ns40:cooperation-conditions>
```
If I use https://emn178.github.io/online-tools/sha256.html, I get hash as a8d887727c7d2bf1ee97e42b828e469e226c62093308a5d820b5a9864f57acbf 
for data under cooperation-conditions section. 

Please suggest if it is correct.

**Comment by pmarinelli** [comment]
Hello Umesh,
I think that the xml fragment is not in canonical form as per exclusive c14n specs (please, anyone tell me if I'm wrong): there is no namespace declaration for prefix ns40, while I would expect such a declaration within the start tag of the cooperation-conditions element.
In general, in order to conclude whether a cooperation-condition subtree is in canonical form or not, the whole iias-get-response document is needed. In your case, I think the missing ns declaration suffices to conclude that it is not.
The hash calculation per-se is straightforward. Trickier is the generation of the canonical form of the cooperation-conditions subtree.
I strongly recommend to use a library for that. 
We are developing our ewp connector in java and we opted for apache santuario for the canonicalization. We select the subtree to canonicalize through XPath.

**Comment by umesh-qs** [comment]
Hi @pmarinelli 
Thanks for responding.

Below is the xml fragment after converting to c14n specs

```
<ns40:cooperation-conditions xmlns:ns40="https://github.com/erasmus-without-paper/ewp-specs-api-iias/blob/stable-v3/endpoints/get-response.xsd">
<ns40:student-traineeship-mobility-spec>
<ns40:sending-hei-id>uni1</ns40:sending-hei-id>
<ns40:receiving-hei-id>uni2</ns40:receiving-hei-id>
<ns40:receiving-academic-year-id>2018/2019</ns40:receiving-academic-year-id>
<ns40:mobilities-per-year>1</ns40:mobilities-per-year>
<ns40:recommended-language-skill>
<ns40:language>en</ns40:language>
<ns40:cefr-level>C1</ns40:cefr-level>
</ns40:recommended-language-skill>
<ns40:isced-f-code>0313</ns40:isced-f-code>
<ns40:total-months>2</ns40:total-months>
<ns40:blended>false</ns40:blended>
</ns40:student-traineeship-mobility-spec>
</ns40:cooperation-conditions>
```

and below is the hash
4f0fc9f8904f63d687873e64e0d1674059aa33955535ebee7ed6e78efb9cd610

Does it look good now?

**Comment by mkurzydlowski** [comment]
Looks good to me. We will update our own EWP connector accordingly as we don't calculate the hash correctly currently.

**Comment by frangarcj** [comment]
If I'm not wrong the c14n form is without an XML namespace name

```<cooperation-conditions xmlns="https://github.com/erasmus-without-paper/ewp-specs-api-iias/blob/stable-v3/endpoints/get-response.xsd"><student-traineeship-mobility-spec><sending-hei-id>uni1</sending-hei-id><receiving-hei-id>uni2</receiving-hei-id><receiving-academic-year-id>2018/2019</receiving-academic-year-id><mobilities-per-year>1</mobilities-per-year><recommended-language-skill><language>en</language><cefr-level>C1</cefr-level></recommended-language-skill><isced-f-code>0313</isced-f-code><total-months>2</total-months><blended>false</blended></student-traineeship-mobility-spec></cooperation-conditions>```

**Comment by umesh-qs** [comment]
From https://www.w3.org/TR/2002/REC-xml-exc-c14n-20020718/. I see the below example.

```
As a simple example of the type of problem that changes in XML context can cause for signatures, consider the following document:

   <n1:elem1 xmlns:n1="http://b.example">
       content
   </n1:elem1>
this is then enveloped in another document:

   <n0:pdu xmlns:n0="http://a.example">
      <n1:elem1 xmlns:n1="http://b.example">
          content
      </n1:elem1>
   </n0:pdu>
The first document above is in canonical form. But assume that document is enveloped as in the second case. The subdocument with elem1 as its apex node can be extracted from this second case with an XPath expression such as:

 (//. | //@* | //namespace::*)[ancestor-or-self::n1:elem1]
The result of applying Canonical XML to the resulting XPath node-set is the following (except for line wrapping to fit this document):

   <n1:elem1 xmlns:n0="http://a.example"
             xmlns:n1="http://b.example">
       content
   </n1:elem1>
Note that the n0 namespace has been included by Canonical XML because it includes namespace context. This change which would break a signature over elem1 based on the first version.
```

**Comment by frangarcj** [comment]
But for example, you are using ns40 as namespace name and maybe my XML manipulation library uses another (eg. ns59) so our hashes for the same version and conditions will be different.

We are using the following function for JAVA with the CooperationConditions class marked as XMLRootElement. The output is with the namespace but without a name.

```java
private String calculateHash() {
        try {
            org.apache.xml.security.Init.init();
            JAXBContext contextObj = JAXBContext.newInstance(IiasGetResponse.Iia.CooperationConditions.class);
            Marshaller marshallerObj = contextObj.createMarshaller();
            StringWriter sw = new StringWriter();
            marshallerObj.marshal(cooperationConditions, sw);
            Canonicalizer canon = Canonicalizer.getInstance(Canonicalizer.ALGO_ID_C14N_EXCL_OMIT_COMMENTS);
            byte[] canonXml = canon.canonicalize(sw.toString().getBytes());
            String canonString = new String(canonXml);
            log.info(canonString);
            
            final byte[] digest = MessageDigest.getInstance("SHA-256").digest(canonXml);            
            return Hex.encodeHexString(digest);
        } catch (Exception e) {
            log.error(e);
        }
        return null;
    }
```

**Comment by umesh-qs** [comment]
It does not matter what namespace prefix is used. You should be calculating the hash as per what data the partner has sent and match that hash. Purpose of hash validation is to make sure that the same hash is calculated by both parties, on the content that is shared in the IIA

**Comment by frangarcj** [comment]
But the hash is calculated over the string characters so if your string contains ns40 and mine doesn't, hashes will be different.

My string to be hashed

```
<cooperation-conditions xmlns="https://github.com/erasmus-without-paper/ewp-specs-api-iias/blob/stable-v4/endpoints/get-response.xsd"><student-studies-mobility-spec><sending-hei-id>demo.usos.edu.pl</sending-hei-id><receiving-hei-id>ual.es</receiving-hei-id><receiving-academic-year-id>2020/2021</receiving-academic-year-id><receiving-academic-year-id>2021/2022</receiving-academic-year-id><receiving-academic-year-id>2022/2023</receiving-academic-year-id><receiving-academic-year-id>2023/2024</receiving-academic-year-id><receiving-academic-year-id>2024/2025</receiving-academic-year-id><receiving-academic-year-id>2025/2026</receiving-academic-year-id><receiving-academic-year-id>2026/2027</receiving-academic-year-id><receiving-academic-year-id>2027/2028</receiving-academic-year-id><mobilities-per-year>5</mobilities-per-year><recommended-language-skill><language>en</language><cefr-level>B2</cefr-level></recommended-language-skill><subject-area><isced-f-code>0410</isced-f-code></subject-area><total-months>6</total-months><blended>false</blended><eqf-level>6</eqf-level></student-studies-mobility-spec><student-studies-mobility-spec><sending-hei-id>ual.es</sending-hei-id><receiving-hei-id>demo.usos.edu.pl</receiving-hei-id><receiving-academic-year-id>2020/2021</receiving-academic-year-id><receiving-academic-year-id>2021/2022</receiving-academic-year-id><receiving-academic-year-id>2022/2023</receiving-academic-year-id><receiving-academic-year-id>2023/2024</receiving-academic-year-id><receiving-academic-year-id>2024/2025</receiving-academic-year-id><receiving-academic-year-id>2025/2026</receiving-academic-year-id><receiving-academic-year-id>2026/2027</receiving-academic-year-id><receiving-academic-year-id>2027/2028</receiving-academic-year-id><mobilities-per-year>4</mobilities-per-year><recommended-language-skill><language>en</language><cefr-level>B1</cefr-level></recommended-language-skill><subject-area><isced-f-code>0410</isced-f-code></subject-area><total-months>5</total-months><blended>false</blended><eqf-level>6</eqf-level></student-studies-mobility-spec></cooperation-conditions>
```

and its hash

```
f192b5470981b688147f94e836ac5cb0c0c7703ef4ee1bf1dadb56933a2d41f7
```

**Comment by fmapeixoto** [comment]
> It does not matter what namespace prefix is used. You should be calculating the hash as per what data the partner has sent and match that hash. Purpose of hash validation is to make sure that the same hash is calculated by both parties, on the content that is shared in the IIA

I understand this sentence and it can be used for that, but I am not quite sure they shouldn't match.
Assuming they shouldn't match, the Approvals API will expose the partner Hash as approved, right? And the partner should expose ours in order for the IIA to be approved?

**Comment by frangarcj** [comment]
But the hash utility is also used to validate the IIA from your partner. If the conditions of the IIA do not generate the hash given in the document, then it is malformed and we can't trust it.

So after we do a GET request of an IIA we must calculate the hash from the cooperation conditions and it must match with the one included in the document. Therefore, both partners must calculate the same hash from the same copy of the IIA.

**Comment by umesh-qs** [comment]
> But the hash is calculated over the string characters so if your string contains ns40 and mine doesn't, hashes will be different.
> 
> My string to be hashed
> 
> ```
> <cooperation-conditions xmlns="https://github.com/erasmus-without-paper/ewp-specs-api-iias/blob/stable-v4/endpoints/get-response.xsd"><student-studies-mobility-spec><sending-hei-id>demo.usos.edu.pl</sending-hei-id><receiving-hei-id>ual.es</receiving-hei-id><receiving-academic-year-id>2020/2021</receiving-academic-year-id><receiving-academic-year-id>2021/2022</receiving-academic-year-id><receiving-academic-year-id>2022/2023</receiving-academic-year-id><receiving-academic-year-id>2023/2024</receiving-academic-year-id><receiving-academic-year-id>2024/2025</receiving-academic-year-id><receiving-academic-year-id>2025/2026</receiving-academic-year-id><receiving-academic-year-id>2026/2027</receiving-academic-year-id><receiving-academic-year-id>2027/2028</receiving-academic-year-id><mobilities-per-year>5</mobilities-per-year><recommended-language-skill><language>en</language><cefr-level>B2</cefr-level></recommended-language-skill><subject-area><isced-f-code>0410</isced-f-code></subject-area><total-months>6</total-months><blended>false</blended><eqf-level>6</eqf-level></student-studies-mobility-spec><student-studies-mobility-spec><sending-hei-id>ual.es</sending-hei-id><receiving-hei-id>demo.usos.edu.pl</receiving-hei-id><receiving-academic-year-id>2020/2021</receiving-academic-year-id><receiving-academic-year-id>2021/2022</receiving-academic-year-id><receiving-academic-year-id>2022/2023</receiving-academic-year-id><receiving-academic-year-id>2023/2024</receiving-academic-year-id><receiving-academic-year-id>2024/2025</receiving-academic-year-id><receiving-academic-year-id>2025/2026</receiving-academic-year-id><receiving-academic-year-id>2026/2027</receiving-academic-year-id><receiving-academic-year-id>2027/2028</receiving-academic-year-id><mobilities-per-year>4</mobilities-per-year><recommended-language-skill><language>en</language><cefr-level>B1</cefr-level></recommended-language-skill><subject-area><isced-f-code>0410</isced-f-code></subject-area><total-months>5</total-months><blended>false</blended><eqf-level>6</eqf-level></student-studies-mobility-spec></cooperation-conditions>
> ```
> 
> and its hash
> 
> ```
> f192b5470981b688147f94e836ac5cb0c0c7703ef4ee1bf1dadb56933a2d41f7
> ```

Are you sure you are not manipulating the XML response when converting to class IiasGetResponse.Iia.CooperationConditions.class?

**Comment by frangarcj** [comment]
I am marshalling an object of that class to a c14n string so there's no manipulation. Just from Object to String.

We have been doing tests with the USOS platform and the University of Salamanca so we think we are doing it correctly.
However, after seeing your XML with the namespace name I had to describe what we are doing for confirmation.

**Comment by umesh-qs** [comment]
> I am marshalling an object of that class to a c14n string so there's no manipulation. Just from Object to String.
> 
> We have been doing tests with the USOS platform and the University of Salamanca so we think we are doing it correctly.
> However, after seeing your XML with the namespace name I had to describe what we are doing for confirmation.

So sw.toString() is exactly same as what is received from the partner?

**Comment by umesh-qs** [comment]
> I am marshalling an object of that class to a c14n string so there's no manipulation. Just from Object to String.
> 
> We have been doing tests with the USOS platform and the University of Salamanca so we think we are doing it correctly.
> However, after seeing your XML with the namespace name I had to describe what we are doing for confirmation.

I don't work on Java. But looking at the code, I think you are converting XML string to a user-defined class and then that class back to a string. If the elements in XML response are not in the same sequence as in your class then it will not generate the XML back in the same sequence and may be name space prefix is also changed

**Comment by pmarinelli** [comment]
Dear All,

As far as we have understood the conditions hash, it has the main purpose to check if changes on an agreement have occurred. 

We ask you what do you think about making the conditions hash reflect changes on data important from IRO member perspective, instead of changes on how such data is serialized.

That is, we think an IRO member is interested in knowing if the updated version of an agreement has changes on the number of mobilities, subject areas, and so on, and not in knowing if it is now served into a new version of the IIA API. 

We think that having an algorithm as less coupled to technicalities as possible would improve the approval process.

Moreover, there is another aspect we would like to point your attention into. EWP allows to expose agreements in multiple versions at the same time. We think it is useful from an interoperability point of view. Do you think that it is appropriate for the IIA Approval API to cover such a use case, for instance by inserting the IIA API version along with the conditions hash? If different IIA API versions produce different conditions hashes, such an extra information would allow the IIA Approval API client know which API version the server has read the agreement from and thus which hash algorithm has to be used to check if it has approved the latest version of the agreement.

**Comment by frangarcj** [comment]
@umesh-qs
1. The XML string/request is first converted to java objects by the JAXB runtime. Other languages maybe can work directly over the received document.
2. To check the cooperation conditions, the elements (objects) are converted to a string, but in c14n form. That is, we are generating an XML string in a standard way that is different from the received document. Therefore, this string has no namespace names, comments or spaces, tabs, line returns. It keeps only the order of the elements.

I agree with @pmarinelli that having the namespace (version of the API) as part of the hash makes maintaining multiple APIs very difficult. If you call the IIA Approval you just receive a hash you don't know the version of.

**Comment by mkurzydlowski** [comment]
> It does not matter what namespace prefix is used. You should be calculating the hash as per what data the partner has sent and match that hash. Purpose of hash validation is to make sure that the same hash is calculated by both parties, on the content that is shared in the IIA

I agree with @umesh-qs in both cases. This is precisely why we need to fix the way we currently calculate the hash.

The way presented by @frangarcj is the current way we calculate hash. It is, sadly, not the way C14N should be applied to the IIA response, as I understand it right now. Although it is an easier way to implement (at least for us) and, in this case, it might even produce the same hash for both partners (but this is not a requirement!).

You are right that the hash will change with every namespace change but it will probably change anyway because of changes to the schema.

Still for keeping approvals in a way that would work indefinitely one would need to use something like signed PDFs. That's a different discussion.

PS. @fmapeixoto, @georgschermann, how are you calculating the hash currently?

**Comment by frangarcj** [comment]
> The way presented by @frangarcj is the current way we calculate hash. It is, sadly, not the way C14N should be applied to the IIA response, as I understand it right now. Although it is an easier way to implement (at least for us) and, in this case, it might even produce the same hash for both partners (but this is not a requirement!).

Why is not the way C14N should be applied? Which is the correct one?

Also, I need to check that the hash given by the partner is correct (if not the iia is malformed), then, why is calculating the same hash not a requirement?

**Comment by mkurzydlowski** [comment]
> Why is not the way C14N should be applied? Which is the correct one?

As @umesh-qs was pointing it out, we are calculating a hash of a subdocument of the IIA get response. That's why we need to take into consideration the namespace prefixes used in that response. I wasn't fully aware of it at the beginning but it is my understanding after further reading of the spec. Still, you might prove me wrong after reading the spec.

> Also, I need to check that the hash given by the partner is correct (if not the iia is malformed), then, why is calculating the same hash not a requirement?

What we want to check is if the hash the partner sends corresponds to the data he is hashing. Neither the hash nor the exact XML representation of the IIA served by us is required to match that of the partner's. I think @kamil-olszewski-uw might try to explain it further.

**Comment by frangarcj** [comment]
We are computing the exclusive c14n correctly, but it is true that there is only one namespace used (the default one) in the responses we are receiving.

```xml
<iias-get-response
    xmlns="https://github.com/erasmus-without-paper/ewp-specs-api-iias/blob/stable-v6/endpoints/get-response.xsd"
    xmlns:c="https://github.com/erasmus-without-paper/ewp-specs-types-contact/tree/stable-v1"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="
        https://github.com/erasmus-without-paper/ewp-specs-api-iias/blob/stable-v6/endpoints/get-response.xsd
        https://raw.githubusercontent.com/erasmus-without-paper/ewp-specs-api-iias/stable-v6/endpoints/get-response.xsd
    "
>
...
```
Using an exclusive c14n it just removes c namespace. However, it's true that nothing restricts the use of prefixes.

> Neither the hash nor the exact XML representation of the IIA served by us is required to match that of the partner's

But I think that we need to add the requirement of having the same hash for both partners if the data is the same. Furthermore, it should not depend on the API version (only if the algorithm changes) or namespace prefixes.

**Comment by mkurzydlowski** [comment]
> We are computing the exclusive c14n correctly, but it is true that there is only one namespace used (the default one) in the responses we are receiving.

That wasn't obvious from the way you are calculating the hash. If you generate the subdocument XML by marshaling only the subelement that there is no way to know for sure that the namespace(s) used in the subdocument have the same prefix names (or lack a prefix) as in the response XML.

> But I think that we need to add the requirement of having the same hash for both partners if the data is the same.

We are not able to enforce this as the data obtained from both systems might have some differences but still may represent valid IIAs that are matching on a business level. @kamil-olszewski-uw will probably explain it better.

> Furthermore, it should not depend on the API version (only if the algorithm changes) or namespace prefixes.

Technically it must change if we stick to this canonicalization algorithm. Also, as I pointed earlier, there is a high probability of fields changing in the hashed subdocument that might change the hash.

Still one might just keep the old API implementation for some time.

I'm still hoping that we might use a different mechanism for "signing" the IIA that would require the sign to live forever, even when the API changes.

**Comment by kamil-olszewski-uw** [comment]
> What we want to check is if the hash the partner sends corresponds to the data he is hashing. Neither the hash nor the exact XML representation of the IIA served by us is required to match that of the partner's. I think @kamil-olszewski-uw might try to explain it further.

If we are going to approve the agreement, our system calls IIAs API get for us. Inside the get response, we get hash and cooperation conditions. Based on the just received cooperation conditions, we compute the hash on our side and check if both hashes match. If so, then we save partner's hash in our system and send the IIA Approval CNR. If the partner sends us an IIA Approval get request, we will send him back his hash in response.

The data of agreement stored in our system is not involved in any stage of this process.

**Comment by georgschermann** [comment]
> PS. fmapeixoto, @georgschermann, how are you calculating the hash currently?

we calculate it nearly in the exact same way as frangarcj
we only use it to check the received IIA not to compare it with our own data, since this will be represented differently most of the time.

**Comment by j-be** [comment]
I currently stumbled over this issue while implementing in at our side (Java with JAX-RS/JAX-B) and I too find the way this is specified highly unhandy. In detail, how I understand it (hard to tell how this would work on some framework out there, so please excuse any short-sightedness):

* To be absolutely sure one would need to perform raw string operations on the raw XML received or to be sent by the API. Any library (even simple DOM parsers) may or may not mangle with it and introduce non-functional changes, like the above mentioned namespaces. They *shouldn't*, but who knows...
* We need this to countercheck data within the same document. In other words: this is explicitly not meant to be a shared key, or any other means of retrieving related documents, or as a means of retrieving the document itself.

So I can't really see why it needs to be as complicated as it is, or what it should be used for. The things I can think of:

* Notice if something changes: Any collision resistant random value would do (see [OLA's `changes-proposal.id`](https://github.com/erasmus-without-paper/ewp-specs-api-omobility-las/blob/3c76b350bc33b3a52acb8673bc958dafba871ef9/endpoints/get-response.xsd#L239-L249))
* Make sure the response was not tampered with: that is handled by signing stuff
* Make sure the data is "correct": the sender will most probably compute the hash over the raw XML string it is about to send. So any fault in the implementation leading to "wrong" values will still result in valid hashes

In other words: I would propose to take one step back, and reevaluate what we are actually trying to achieve here, and if this could maybe be achieved in an easier way (e.g. like the nonce used in OLA's) rather than operations on raw strings and hashing that.

**Comment by ctu-developers** [comment]
Hello everyone.

I think the second part of the discussion in this ticket belongs to #47. We reported a problem with using the XML-C14N function to calculate conditional hashes more than a year ago, when we implemented IIAs V4. It's a long time. The documentation was poor. I agree with #53 we thought about it too. If I remember well...

For the process described here -- https://github.com/erasmus-without-paper/ewp-specs-api-iias/issues/48#issuecomment-815836903

> > What we want to check is if the hash the partner sends corresponds to the data he is hashing. Neither the hash nor the exact XML representation of the IIA served by us is required to match that of the partner's. I think @kamil-olszewski-uw might try to explain it further.
> 
> If we are going to approve the agreement, our system calls IIAs API get for us. Inside the get response, we get hash and cooperation conditions. Based on the just received cooperation conditions, we compute the hash on our side and check if both hashes match. If so, then we save partner's hash in our system and send the IIA Approval CNR. If the partner sends us an IIA Approval get request, we will send him back his hash in response.
> 
> The data of agreement stored in our system is not involved in any stage of this process.

This is true only in one case - both sides use the same prefixes like _c_, _ewp_, _trd_ (from documentation). Using **XML-C14N**, we are **not able** to compute an **identical hash** from the partner data. This function does not guarantee an exact form for generating the identical hash. See XML-C14N specification -- https://www.w3.org/TR/xml-exc-c14n

What prevents me from parsing partner's data into my DOM tree and using my prefixes? Nothing. Another case. So if the partner uses other namespace names, the checksum will change without changing anything in the content. Doesn't the purpose of the "condition hash" to verify that the data hasn't changed? 

Yes, the problem can be circumvented by going through the whole process - building a DOM, generating XML with my namespace and comparing the checksum with the version I sent in the previous round. So I am able to check that the partner in the document has not changed anything (except the namespace name). But it's a little annoying. It wouldn't matter yet, but somewhere only those checksums are sent without the content they "secure" and so the control depends only on my original data. That is the bigger problem, I think.

**Comment by j-be** [comment]
@ctu-developers I'm a bit confused now. As far as I understood, the `condition-hash` of the partners are not required to be the same in their respective representation of any given IIA. This is because the lists (e.g. `student-studies-mobility-spec`) are not required to have a deterministic ordering, only ["consistent"](https://github.com/erasmus-without-paper/ewp-specs-api-iias/blob/55aa7276dbc54a01397cfa3040f053d5c5d926b4/endpoints/get-response.xsd#L193).

 See the examples: [A's XML](https://github.com/erasmus-without-paper/ewp-specs-api-iias/blob/55aa7276dbc54a01397cfa3040f053d5c5d926b4/example-scenario/01-A-get-response.part.xml#L31) has a different `conditions-hash` than [B's XML](https://github.com/erasmus-without-paper/ewp-specs-api-iias/blob/55aa7276dbc54a01397cfa3040f053d5c5d926b4/example-scenario/03-B-get-response.part.xml#L32). This is due to both listing the conditions where they themselves are `sending-hei` first, effectively reversing the list.

That is why I don't understand the strict requirement on the hash: I don't see any benefit in being able to check the hash on my side. Only thing I care about is that the one I am seeing now is the same as the one I saw before. And for that this procedure seems way overengineered to me. Any nonce (i.e. a simple version counter would do) is plenty. Or am I missing something?

**Comment by georgschermann** [comment]
you can see most of the discussion regarding the hash here [#30](https://github.com/erasmus-without-paper/ewp-specs-api-iias/issues/30).
There has also been a discussion to use status flags or version numbers or the like.
Hash was preferred by some partners,

**Comment by j-be** [comment]
@georgschermann I see, thanks for the link.

So is my understanding correct, that a hash was chosen, rather than a simple nonce due to lack of trust between HEIs? In other words: If I blindly trust PartnerHEI, I gain no benefit in checking the hash, right?

Follwoing that thought: is it true that the hash is not intended to be used for anything else but establishing trust, i.e. assuring an IIA does not change unilaterally and unnoticed after it has been approved by the partner? In particular, it is explicitly not a means of retrieving, linking, or referencing (except for Approval API) IIAs, right?

**Comment by jiripetrzelka** [comment]
What is the benefit of including namespaces in the hashed cooperation conditions? Aside from conforming to the W3C specification, it just breaks the hash whenever a new version of the API is released any thereby the namespace is changed, even if there has been no change in the part of the XML that is being hashed. Isn't this unnecessary?

**Comment by georgschermann** [comment]
The hash was chosen to be able to check/verify at a later time if the agreement is still the one which has been approved, since you could store the signed approval response. You are also always able to reproduce the same conditions you had previously even when namespaces etc. from the live API changed in the meantime. So you would always be able to prove, that your IIAs had been approved.

As a partner you would not recognize an IIA as approved by you when there has been a namespace / hash change. But you could also store the IIA GET response to prove that this was the version you had approved.

In our implementation we don't require/use the approval for anything, since we did our implementation before the approval API was introduced and we think of the signed-fields on the IIA as the more important information. But there are some partners which implemented the approval as an integral part of their processes, so we had to add several auto-approvals in our process to be able to exchange data with these partners.

The implementations of several partners differ greatly regarding the approval to cover their internal processes and it is still under discussion how to solve these issues which became visible during the last weeks and months.

If you don't rely on being able to prove an earlier approval in your software you can omit the whole thing and auto approve e.g. on import/signing/etc. or wait for the different partners to agree on a spec change e.g. towards a nonce / approval-uid / namespace-free hash calculation / etc.

**Comment by sascoms** [comment]
I think adding a namespace to the coop parts before hashing is only adding an extra job/task and it only makes the already complicated IIA process more complicated regarding there are many different scenarios, different implementations etc.

Whatever the decision is or will be, in my opinion, it should be (if not yet) decided very very soon whether we need to add a namespace or not and also be documented on the API specs. 

As some of you know or already experience, this is causing IIA daha exchange problems between providers. (as some adds ns and some not).

**Comment by muratyuceer** [comment]
When i first read the document, I was thinking hashes between local copies of partners must be equal and I was thinking of comparing them to realize that we are equal before approve.

After read issue comments and learn XML-C14N calculation, my thoughts changed to;

When I need to check a hash from a partner I have to calculate the hash again from the partner raw xml response (partner "get" api result (because version or namespace prefix could be different)) and compare it. In this case I will not use this calculation to understand the difference but just to confirm that we are using the same algorithm and store it if you want. (Also if i skip that I don't know what to lose)

To understand the difference I always have to calculate a hash from the bind api response to my c# class and compare it from my local hash.

This is what i'm thinking of doing

**Comment by j-be** [comment]
@muratyuceer As far as I understood it, the respective hashes are not expected to match - ever. This can have a variety of reasons (i.e. pretty-printed XML vs. all in one line), the most obvious being, that the spec requires the ordering of the elements within the lists in `cooperation-conditions` to be "consistent", but not deterministic.

To better see what I mean: check out the 2 get responses in [the example](https://github.com/erasmus-without-paper/ewp-specs-api-iias/tree/stable-v6/example-scenario): Both `A` and `B` list the entry, where they themselves are "sending HEI" first, effectively reversing the list, and thus changing the hash.

So as far as I understood: the only way for me to countercheck the hash of a received IIA is to perform operations on the raw XML string, which depending on the framework (Java JAX-RS with JAX-B here) may become particularly tricky - especially the "remove contacts" part of it.

**Comment by pmarinelli** [comment]
@j-be, we use XPath to extract the node set to canonicalize. Of course the XPath expression is applied to the octect stream as received by the partner. The use of XPath sounds to me as the safest choice, as it is also referred to by the exc-C14n specs (https://www.w3.org/TR/xml-exc-c14n/), which states: "The exclusive canonical form of a document subset is a physical representation of the XPath node-set, as an octet sequence, produced by the method described in this specification."

**Comment by j-be** [comment]
@pmarinelli thanks, seems like I need to look at XPath expressions (I always avoided them as I really dislike the concept of workling on raw XML) :smile: .

Anyway, after thinking about this a bit more, I think I am more confused than ever. Take the following (as far as I can tell XML-C14N canonical) XML (some stuff omitted for brevity):

```xml
<cooperation-conditions>
  <student-studies-mobility-spec>
    <sending-hei-id>tuwien.ac.at</sending-hei-id>
    <sending-contact>
      <contact-name>Doesn't really matter</contact-name>
    </sending-contact>
  </student-studies-mobility-spec>
</cooperation-conditions>
```

After [*excluding* `sending-contact` and `receiving-contact` subelements](https://github.com/erasmus-without-paper/ewp-specs-api-iias/blob/stable-v6/endpoints/get-response.xsd#L236-L237) this leaves me with (as far as I can tell still XML-C14N canonical):

```xml
<cooperation-conditions>
  <student-studies-mobility-spec>
    <sending-hei-id>tuwien.ac.at</sending-hei-id>
    
  </student-studies-mobility-spec>
</cooperation-conditions>
```

Note, that there are four ` ` (aka. spaces) in the apparently empty forth line.

In other words: I have to effectively remove the following regex from the XML (assuming no comments here, else it gets really tricky): `<sending-contact>.*</sending-contact>`, but I must not touch any white-space character preceding or following that part. Is that correct?

**Comment by pmarinelli** [comment]
@j-be Yes, it is correct (at least for me). The ewp specs require us to remove certain elements, not the whitespace characters possibly surrounding them.

**Comment by sascoms** [comment]
We are talking on how to hash and how to check hashes etc.

Is it only me who thinks this is rather very complicated and needs simplification or a better solution?

The missing partner IIA IDs, the hash calculation, hash checks, approvals, problems caused by master-master principle, duplicates, etc.

Why do we all need to find many turnarounds, creating magic tricks to fix these gaps, problems, issues?

Don't we need a new version in which IIA workflow is redesigned especially when now more developers and providers are involved and all have experienced the current problems during implementation and data exchange?

**Comment by j-be** [comment]
@sascoms I mostly agree, though I think a lot of this *could* be fixed by adding:

* a deterministic and unique way to calculate the hash (i.e. any 2 copies of an IIA hash to the same value), so the hash can be used as a key, or even an ID, and thus
* allow to search for said hash

This would solve:

* complicated hashing, as it could be moved to the actual content, rather than the representation thereof (aka. XML)
* finding the respective IIA at the partner HEI
* approval is implicit if both HEIs have the same hash

It would even allow for changes in the schema without invalidating everything, as a second, or third hash can be added (though this may not scale beyond a couple of versions).

This could be achieved by adding:

* deterministic sorting to the list
* a hashing scheme, that for the same content (as in "what we agreed on") always returns the same hash

**Comment by pmarinelli** [comment]
@j-be just for sake of clarity.The xpath expression we use excludes by itself the sending and receiving contacts.That is, the node set selected by the xpath expression is ready to be canonicalized, without any further processing. I feel confortable with such an approach as it delegates the handling of all the xml peculiarities to an engine that is specifically designed to deal with them.

**Comment by jiripetrzelka** [comment]
During the implementation of the conditions-hash, we have stumbled upon the `&#xD;` entity in the Mobility Online implementation. Since the conditions-hash did not match, I tried to replace the entity by the newline character and voila the conditions hash started to match. But now I don't know if we are really supposed to convert entities into real characters and only then compute the hash, or if this is a glitch on the part of Mobility Online and the conditions-hash should be computed directly from the entity. Can someone please clarify?

**Comment by jiripetrzelka** [comment]
Would it be possible to add the requirement to the specification that the string being canonized mustn't contain any white spaces between elements so that we are not forced to do raw string operations on the incoming XML to preserve every space and tab in the string, as j-be pointed out in https://github.com/erasmus-without-paper/ewp-specs-api-iias/issues/48#issuecomment-881471580 ?

The Canonical XML Version 2.0 does not imply that spaces be removed:
https://www.w3.org/TR/xml-c14n2/#sec-Requirements-Robustness
![image](https://user-images.githubusercontent.com/1330161/132383848-1d18ced3-b87f-4cdc-9a1d-2ba3641a57d4.png)

If I understand it correctly it means that if the partner adds any specific indentation to the XML then we cannot use any XML library for processing unless it is able to preserve the exact formatting, which I doubt any library can do.

**Comment by janinamincer-daszkiewicz** [comment]
For me it looks that even if some development teams have somehow solved the problem of hash calculation and can exchange IIAs betwen different own installations the problem is not yet solved on a global level and we still try to cope by tightining the requirements expressed in the specification. Which will help in only a short term.

What if we took another approach?

The role of hash is to give a partner whose IIA is being approved some evidence that we approve what has been sent to us in the last IIA get. The advantage of such proof is that hash is a relatively short string, easy to store and compare. Unfortunately it is hard to calculate (in a unified way on a global level). Also it it more difficult to use it in court as we need an accompying element showing that exactly this hash has been sent to us by the partner. Showing system logs in court may not be a good idea.

There is another option. Let's use a whole XML element with IIA (object, document) as such proof. Yes, as a string it is longer than hash but nowadays who cares (still much shorter than PDF equivalent). You may compare it as a string and also you can compare it as an object, with an internal structure, element by element. It is also much more transparent than the hash of this XML element. In particular if strings do not match you can easily find out why, which subelement has changed. May be only contact details and you do not care and will not ask for another round of approvals?

There is also extra option. You can add signature. In fact there are three options and the decision is more technical than business oriented: use document with embedded signature, or signature with embedded document, or document and signature as two separate elements. Signature is a much better court proof than hash (easier to handle). You can show it to the end user in the interface and quite possible that the IRO staff will prefer it over hash.

Summary
Now in IIA get we obtain IIA in XML format and its hash.
Let's eliminate hash.
In IIA approval, instead of sending the obtained hash we would send the obtained IIA in XML format.  The option to consider would be to add signature. Hash calculation would be eliminated from the process.

Tell us your opinion.

**Comment by j-be** [comment]
My two cents: As it is right now, we dicided not to check the partners hash, as falsifying it would need susbtantial criminal intent, which we trust pur partners don't have. We interpret it as a nonce an candidate key for approval, nothing more.

**Comment by georgschermann** [comment]
Currently I don't have an opinion on this, just wanted to mention, that signatures are already present in all requests and responses due to httpsignature authentication and the fact that tls server auth has been dropped by most providers.

@j-be same here

**Comment by umesh-qs** [comment]
I am not sure what purpose is proposed signature method serving instead of hash is serving except for the new legal angle that is being brought into. Even then I am not sure how it can be used in courts. If some more details can be provided on this from the EUF perspective, then it will help.

The current problem with calculating hash with its definition. If we can agree on a common format, like creating a single line string without any spaces and new line characters between tags and removing all comments, namespaces and prefixes, then the hash will work fine.

Also, most of the developers here can live/are living without the hash functionality. So better drop the hash comparison in case we cannot come up with a common format

**Comment by BavoNootaert** [comment]
I agree with @georgschermann and @umesh-qs: we already have signatures in the http headers, and hashes should work fine if we all just follow clear specifications. E.g.. compute the hash using the canonicalization on the data that is actually sent. Using XML Signatures involves canonicalization, so it could lead to similar problems. Moreover, the XML Signature specification is more complex, contains 'recommendations' and 'should's, and warnings like this:
> Alternatives to the REQUIRED canonicalization algorithms (section 6.5), such as Canonical XML with Comments (section 6.5.1) or a minimal canonicalization (such as CRLF and charset normalization) , may be explicitly specified but are NOT REQUIRED. Consequently, their use may not interoperate with other applications that do not support the specified algorithm

So the problems may be even worse.

**Comment by mkurzydlowski** [comment]
@BavoNootaert, the way XML signature is prepared doesn't have to be strict as it won't be recalculated by the partner (as it has place with hashes currently).

The XML signature is not intended to "replace" hash computation, it's solely to be a prove and a good one.

What was suggested as a remedy for hash computation is just storing the whole XML, rather than a hash.

**Comment by BavoNootaert** [comment]
How can you be sure it's a proof if you don't check it is correct first?

As for storing the whole XML, I think one of the problems with computing the hash is that some frameworks make it hard to access the orginal request or do something with the response after hashing. So it cannot be assumed the XML that will be included in the approval is that what was sent by the partner. It may even be restructured to match the database structure of the approving partner (fields truncated or removed, reordered, whitespace added...). In that case it will not be an easy comparison at all. Checking it would require user intervention whereas checking hashes is done automatically. That is a _huge_ change.

**Comment by mkurzydlowski** [comment]
What do you mean by "correct"?
- The verification of a signature will be handled by the library.
- The XML should correspond to the partner response.

The XML doesn't need to correspond byte by byte to a server response. It needs to represent the same data. Comparing two XML representations should also be handled by libraries (XMLunit for example but this was just the first search result).

**Comment by BavoNootaert** [comment]
Of course one would use libraries as much as possible. The point is: the hash is effectively a reference to a response received earlier (in some of the comments above it is treated as a nonce.) That is relatively straightforward. 

The entire XML is harder:
- Extracting the response as it was received by the partner is awkward with many frameworks, as it is parsed to some object before it is handled by the application itself. So if you include it in the approval response, you have the same awkwardness as with hashes. That part isn't solved. 
- Upon receiving the approval response, you have to compare that xml with you own data. Instead of a simple string comparison, you now have to resort to libraries to  help  you. For which you still have to specifiy what is a significant difference and what is not. (If that is at all possible, given that partners may store the same IIA differently.) So you have made things more complex.

I think the legal argument (that it is a better proof) should be moved to a different issue, as it solves a different problem, and leads to other questions.

**Comment by demilatof** [comment]
If some one is interested, I tried here https://github.com/erasmus-without-paper/ewp-specs-api-iias/issues/72#issue-1069661511 to explain a different approach to the problem; this approach can produce two possible solutions:

1) sharing the same algorithm or exposing the exact string on a single line to compute for hash, between tags `<toHash>` and `</toHash>`
2) indicating that there is no need to compute again the hash received, just saving it together with IIA_id, and writing it in the approval. Every developer will choose how to bind iia_id, hash and data: storing XML or single fields.

**Comment by mkurzydlowski** [comment]
Returning to the issue of emulating signature with a hash.

For HEI A to be able to acquire a "signature" on A's IIA it currently needs to:

1. Store B's IIA approval of A's IIA. Otherwise if B looses, changes or can't temporarily serve the IIA approval, then A looses its signature.

What's more A has to store B's IIA approval in a way that can be used as a proof of B's will. Currently B would need to implement HTTP signature and A would need to store the data being transmitted. Also B's server key needs to be kept. This key needs to emulate the signing party even if an actual person is a better fit in this scenario.

2. Obtain and store the exact format of it's own IIA being transmitted to B. This can be cumbersome for many implementers and strictly depends on the data not being changed between B's request for A's IIA and B's approval.

For a "signature" to be emulated we need both IIA XML and "signed" IIA approval XML, and also be sure that hashes are equal.

It should be noted that such "signature" would not be subject to IIA data and version changes, as both IIA and approval would be stored based on the moment of signing.

It should also be noted that what has been described above is in fact a simulation of XML signature but a very cumbersome and error prone one. For an XML signature to take place:

1. B has to take A's IIA XML and sign it (all of it - no XML operations needed) by an XML signature library.
2. A has to store this signed XML.

Such signed XML is in fact a well defined proof and it additionally keeps the exact information about the signing party.

**Comment by janinamincer-daszkiewicz** [comment]
Some changes in the IIA and IIA Approval APIs will take place by the end of 2022. It makes sense to group them to make the change in the major number of the APIs once.

Have a look at arguments listed in https://github.com/erasmus-without-paper/ewp-specs-api-iias/issues/48#issuecomment-1004723182.

During the Infrastructure Forum meeting on 2022-11-16 the providers will vote if we want to replace hash with XML signature as proposed by UWarsaw.

**Comment by demilatof** [comment]
My bad, I don't completely understand the advantage of XML signature respects the hash code, whilst it could offer similar issues in hash computation.
Since the specifications give importance to the cooperation conditions hash code, I thought that there was a reason: "signing" the essential part of an IIA and allowing changes elsewhere.

If introducing XML signature is for saving a copy of the XML signed by the partner (as a proof), the same could be achieved saving our XML and the Approval Response where there is the hash code.
To be noticed that when B takes A's IIA XML, B could even modify the XML before signing it (intentionally or because of a library or an intermediary). Therefore, when A downloads the signed copy, it has to check that B has signed what A previously sent. We have to take care of all of these issues, keeping in mind that the two documents could contain the same information, but an empty space or a carriage return would make them different, breaking the XML signature.

**Comment by umesh-qs** [comment]
There is no advantage of XML Signature over current hash in current context. This just seems a fancy proposal with no real need.
Not in favor of it for the proposed use case.

**Comment by jiripetrzelka** [comment]
By XML Signatures you mean what we already have in ToRs, i.e. this element?:

https://github.com/erasmus-without-paper/ewp-specs-api-imobility-tors/blob/d051d8a66b7fa86624590109529669832a31d4d4/endpoints/get-response-example.xml#L61-L94

Do you suggest to encrypt the entire IIA contents and not only the cooperation-conditions element?

Do you suggest to encrypt the raw xml or first canonize the xml?

Which/whose private key do you suggest to use for encryption?

How would the approvals work? Would partners still approve some kind of digest/hash?

What is the main added value of introducing this change?

**Comment by mkurzydlowski** [comment]
@demilatof, @umesh-qs and @jiripetrzelka asked about added value of XML Signature solution but I'm not really able to add anything more than what I tried to summarize in my [previous comment](https://github.com/erasmus-without-paper/ewp-specs-api-iias/issues/48#issuecomment-1004723182). This solution gives the simplest way (calling XML signing library) to obtain a proper proof of partner's will (properly signed XML with credentials). It is intended to replace the current solution that proved to be very cumbersome and misleading.

> Do you suggest to encrypt the entire IIA contents and not only the cooperation-conditions element?

Entire IIA.

> Do you suggest to encrypt the raw xml or first canonize the xml?

Raw XML.

> Which/whose private key do you suggest to use for encryption?

We might decide to impose whose credentials should be used for signing but there is a possibility, in this solution, to acknowledge that an agreement is signed be an actual person.

> How would the approvals work? Would partners still approve some kind of digest/hash?

By returning a signed XML with the partner's IIA. The side that gets and approval may (but is not obliged to) check if the signed XML is in fact what it supposed to be. But I wouldn't stress on that, as the IIA might changed in time (data or IIA format) but what is crucial is the exact IIA "state" that has been signed and the fact of signing.

> What is the main added value of introducing this change?

As I wrote in my previous comment, we want to have a proper proof and a simple one to obtain.

**Comment by umesh-qs** [comment]
@mkurzydlowski 
I am not sure who wants to have a proper proof and for what? What purpose is not severed by current hash process?

How is XML signature a better proof as compared to hash?
How is the XML signature process simpler as compared to current process?

**Comment by janinamincer-daszkiewicz** [comment]
> I am not sure who wants to have a proper proof and for what? 

This argument we have heard during analysis from various IROs. Some told us that occasionally they have to go to court and need some proof. This was the main reason the idea to have hash was born.

**Comment by sascoms** [comment]
As well as I remember, this proposal/solution was discussed last year and nobody was in favor of it from the providers.

And we now found it on the desk again. Why? 

Not having a better solution does not mean we need/have to accept a bad or complicated solution.

Better, continue with the current implementation until we find a better one.

**Comment by umesh-qs** [comment]
> > I am not sure who wants to have a proper proof and for what?
> 
> This argument we have heard during analysis from various IROs. Some told us that occasionally they have to go to court and need some proof. This was the main reason the idea to have hash was born.

@janinamincer-daszkiewicz  I assume before proposing this new process as a legally binding proof, you might have checked and confirmed this with legal expert. Please confirm.

**Comment by janinamincer-daszkiewicz** [comment]
> @janinamincer-daszkiewicz I assume before proposing this new process as a legally binding proof, you might have checked and confirmed this with legal expert. Please confirm.

No. Are you suggesting that we do not need hashes at all or that neither the current solution not the proposed one will solve the legal issues?

**Comment by umesh-qs** [comment]
> > @janinamincer-daszkiewicz I assume before proposing this new process as a legally binding proof, you might have checked and confirmed this with legal expert. Please confirm.
> 
> No. Are you suggesting that we do not need hashes at all or that neither the current solution not the proposed one will solve the legal issues?

I am amused with this question. You and Michal (not sure if this is coming from EWP Consortia) are suggesting that the reason for this proposal is for providing valid legal proof, because some IROs have asked for it. And we are discussing that. If the new complex solution is not solving that then why propose it?

**Comment by janinamincer-daszkiewicz** [comment]
> I am amused with this question. You and Michal (not sure if this is coming from EWP Consortia) are suggesting that the reason for this proposal is for providing valid legal proof, because some IROs have asked for it.

Yes. The busines requirement was to provide a legal proof. Yes, the request came from Business Process Owners, meaning IROs.

> And we are discussing that. If the new complex solution is not solving that then why propose it?

I assume that we are not discussing the business need but the technical solution. We have implemented hashes. We discuss other option, HTTP signatures. Some of us argue that it is more complex, some not. We are gathering all arguments here to be ready for making the final decision and voting.

**Comment by umesh-qs** [comment]
> > I am amused with this question. You and Michal (not sure if this is coming from EWP Consortia) are suggesting that the reason for this proposal is for providing valid legal proof, because some IROs have asked for it.
> 
> Yes. The busines requirement was to provide a legal proof. Yes, the request came from Business Process Owners, meaning IROs.
> 
> > And we are discussing that. If the new complex solution is not solving that then why propose it?
> 
> I assume that we are not discussing the business need but the technical solution. We have implemented hashes. We discuss other option, HTTP signatures. Some of us argue that it is more complex, some not. We are gathering all arguments here to be ready for making the final decision and voting.

I don't agree. We are discussing a technical solution to a business problem. And since the trigger for this proposed technical solution is for providing a legally binding proof, a natural question is "Can this new process provide legally binding proof?". And the proposer(s) is not ready to answer this, but still wants a discussion.

**Comment by janinamincer-daszkiewicz** [comment]
> And since the trigger for this proposed technical solution is for providing a legally binding proof, a natural question is "Can this new process provide legally binding proof?". And the proposer(s) is not ready to answer this, but still wants a discussion.

The same question can be asked about hashes and the answer would be the same.

Do you suggest that discussing various technical options does not make sense?

**Comment by umesh-qs** [comment]
Really? Aren't we already discussing this?
Since you are the strong proposer and support of this new solution, you need to have answers to the valid questions. And if there are no answers then it means that proposal is weak.

I am repeating again. Whole premise of this new proposal is legal validity of the approved agreements. If that is not happening then take is back, work on it and come back. But lets close it for now. I wouldn't even consider it for a vote at this stage.

**Comment by jiripetrzelka** [comment]
If I understand it correctly, the main benefit of introducing XML signatures would be increased security. Currently, in case of a dispute, I would show the log containing HTTP communication with partner IIA Approval endpoint and the log containing HTTP communication with my IIA Get endpoint. I would then show that conditions-hashes match and consider it to be the proof of what the partner has approved.

In which case this would not be enough? Well, if anybody with fraudulent intentions could get to the logs and modify them. They could modify anything except HTTP headers signed by partner using Http signature.

By proposing to start signing the entire IIA and not only Http headers, you suggest that the current mechanism is not secure enough. But I would argue that it is not less secure than what coordinators have been using so far - that is, scanned PDFs with pen signatures. Is forging a pen signature harder than forging HTTP communication logs? I wouldn't say so.

@janinamincer-daszkiewicz You mention that Business Process Owners asked for this. Did they mention any legal dispute concerning IIAs in the past in which a pen signature was the weak point? Or have they mentioned that their national or EU legislation requires this level of security or that this kind of legislation is to be introduced in the next 2 years?

@mkurzydlowski Do I understand it correctly that instead of approving a hash or digest, the partner A would take the entire IIA of partner B, sign it with their private key and then expose the entire signature element in their approval endpoint, i.e. something like this?:
https://github.com/erasmus-without-paper/ewp-specs-api-imobility-tors/blob/d051d8a66b7fa86624590109529669832a31d4d4/endpoints/get-response-example.xml#L61-L94

**Comment by jiripetrzelka** [comment]
I have probably made a mistake in my previous comment because Http signature contains message digest on input. So if I am correct, introducing XML signatures is not about increased security but just about a more standard solution. And with the potential to start using personal certificates for signatures. But is it something we need/want?

@mkurzydlowski If the entire IIA would be signed and not just the cooperation-conditions, wouldn't it bring some new problems? What if I change just a contact person? The counterparty will need to inspect the my IIA and find out whether the changes made need a new signature? Wouldn't this complicate the process?

**Comment by demilatof** [comment]
I totally agree with @umesh-qs 

@janinamincer-daszkiewicz  
>  _I assume that we are not discussing the business need but the technical solution._

I don't understand how could be possible discussing the technical solution without fully understanding the business needs.
What kind of legal proof they need? What is the specification of this needs? What analysis have we made?

@mkurzydlowski you said: 
> _This solution gives the simplest way (calling XML signing library) to obtain a proper proof of partner's will (properly signed XML with credentials)_

But on what basis you state that is the simplest way to obtain a proper proof? 
A legal proof must be strong or it is not a legal proof.
And the XML signature doesn't appear to be accepted, not in Italy.
But even the Europe requires a different type of signature; not an XML signature, but XAdES (XML Advanced Electronic Signatures).
This point arises a question: is the choice of XML Signature supported by jurists or lawyers?
I have the doubt that the XML signature will be a solution that complicates everything to solve nothing, because "**_courts are not obliged to accept XAdES-based electronic signatures as evidence in their proceedings; at least in EU, this is compulsory only for "qualified" signatures. A "qualified electronic signature" needs to be doted with a digital certificate, encrypted by a security signature creation device, and the identity of the owner of this signing-certificate must have been verified according to the "high" assurance level of the eIDAS regulation_**"
https://en.wikipedia.org/wiki/XAdES

In order to have a legal proof that is valid in Europe we cannot experiment new solutions that bring only new problems.
And I'm conscious that implementing XML Signature is not so simple as represented; the digital signature requires always a great effort to solve several questions and a deep analysis that cannot be concluded in a couple of months.

For example: what XML Signature? Enveloped, Enveloping or detached signature?
If we choice XAdES, what profile? XAdES-B-B, XAdES-B-T, XAdES-B-LT, XAdES-B-LTA?
What about public/private key, CA and so on? And PKI certificate contents, certificate ordering, certificate revocation and CRL management?
Who will responsible to put the signature? If a human operator, will he/she accept such a duty? 

You also suggest to encrypt the "Raw XML"; but what you mean with "raw XML"? We have already seen that the "RAW cooperation conditions" element was not so "RAW" and unique.
And XML Signature requires canonicalization; if we start using a library, avoiding its requirement we don't make a good use of it.

In my opinion the great problem that underlies the legal proof issue is the master-master model and the distributed architecture.
With a central repository, a single portal and a single language implemented (English) most problems would disappear. 
Everyone makes the rules for incoming student from a foreign Institute, that if accepts put a flag on the proposal.
A makes the rule for B's incoming students and A just accepts the rules created by B on the unique platform. And vice versa.
No more than this; after that web services could be used to export data from the central repository to each internal system to conform the internal processes.
Here instead is everything so overcomplicated...

**Comment by pleys** [comment]
> "Can this new process provide legally binding proof?". And the proposer(s) is not ready to answer this, but still wants a discussion.

On the request from the European Commission we included in the mandatory business requirements:

For the time being and until eSignature becomes a function available in EWP, approval by both parties in the EWP network is considered as the equivalent of a digital signature confirming institutional commitment, provided the institutional legal representative has given an internal mandate. If necessary, due to local rules or regulations, a legal representative can sign Inter-institutional agreements on top of the EWP approval outside the network. In such exceptional cases, HEIs are encouraged to sign digitally and in full compliance with eIDAS legislation.

It will also be added to all official EC documentation about IIAs.

**Comment by demilatof** [comment]
@pleys 
> On the request from the European Commission we included in the mandatory business requirements:
> 
> For the time being and until eSignature becomes a function available in EWP,

I think you have included in the mandatory business requirements a statement too generic.
What kind of eSignature should be available in EWP to satisfy the European Commission' request?

1. electronic signature (natural person) or electronic seal (legal person)?
2. simple, advanced or qualified? Because "**_something as simple as writing your name under an e-mail might constitute an electronic signature_**" but if we want a qualified signature it must be "_**created by a qualified signature creation device (QSCD)**_" and "_**based on a qualified certificate for electronic signatures.**_" (similar for electronic seal)

"**_While different levels of electronic signatures may be appropriate in different contexts, only qualified electronic signatures are explicitly recognized to have the equivalent legal effect of hand-written signatures all over EU Member States._**"

To have almost an Advanced electronic signature/seal (AdES) I think we have to consider XML AdES (XAdES) and not simply XML Signature.
Therefore, to avoid misunderstanding, we should clarify if when we say "XML Signature" we mean XML DSig or XML AdES.
Because in my opinion the simple XML Signature (DSig) doesn't rise up the level from simple to advanced, voiding every effort

https://ec.europa.eu/digital-building-blocks/wikis/display/DIGITAL/eSignature+FAQ

**Comment by umesh-qs** [comment]
@pleys does EC considers "XML Signature" as a concrete legal proof? If so please let us know from whom in EC is this requirement coming from, so that we can directly talk to them and present our case.

**Comment by janinamincer-daszkiewicz** [comment]
This topic is continued in https://github.com/erasmus-without-paper/ewp-specs-api-iias/issues/109.
